### PR TITLE
[client] Send warning event on missing required fields 

### PR DESCRIPTION
**Pull request by ZacharieTFR**
### Description
Trigger a "VAST-warning" if any required attributes or subelements are missing in any element.
According to https://iabtechlab.com/wp-content/uploads/2018/11/VAST4.1-final-Nov-8-2018.pdf

 ### Type
- [x] Breaking change
- [x] Enhancement
- [ ] Fix
- [ ] Documentation
- [ ] Tooling

### CODE REVIEW

File: src/parser/ad_parser.js [hunk 1]
```diff
@@ -27,29 +28,26 @@ export function parseAd(adElement) {
 
     parserUtils.copyNodeAttribute('id', adElement, adTypeElement);
     parserUtils.copyNodeAttribute('sequence', adElement, adTypeElement);
-
-    if (adTypeElement.nodeName === 'Wrapper') {
-      return parseWrapper(adTypeElement);
-    } else if (adTypeElement.nodeName === 'InLine') {
-      return parseInLine(adTypeElement);
-    }
+    return parseAdElement(adTypeElement, emit);
```
Comment (kobawan) [inline src/parser/ad_parser.js]: I'm not sure what we gained by refactoring this part. Previously parsing inlines and wrappers had a more separate logic, and now parseAdElement does everything. To me it was clearer how inlines and wrappers were parsed previously. Let's discuss this :)

File: src/parser/creative_companion_parser.js [hunk 2]
```diff
@@ -41,10 +41,12 @@ export function parseCreativeCompanion(creativeElement, creativeAttributes) {
         .childrenByName(companionResource, 'StaticResource')
         .reduce((urls, resource) => {
           const url = parserUtils.parseNodeText(resource);
+          const creativeType = resource.getAttribute('creativeType') || null;
```
Comment (kobawan) [inline src/parser/creative_companion_parser.js]: there was no need for this change, now `creativeType` constant is created on every loop of reduce

File: src/parser/creative_linear_parser.js [hunk 3]
```diff
@@ -18,10 +18,10 @@ export function parseCreativeLinear(creativeElement, creativeAttributes) {
   let offset;
   const creative = new CreativeLinear(creativeAttributes);
 
+  const durationElement = parserUtils.childByName(creativeElement, 'Duration');
```
Comment (kobawan) [inline src/parser/creative_linear_parser.js]: I would undo these changes in the file as well, as they are not needed anymore

File: src/util/requiredValues.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    subElements: ['MediaFile']
+  },
+  MediaFile: {
+    attributes: ['delivery', 'type', 'width', 'height']
+  },
+  Mezzanine: {
+    attributes: ['delivery', 'type', 'width', 'height']
+  },
+  NonLinear: {
+    inLineRessources: ['StaticResource', 'IFrameResource', 'HTMLResource'],
```
Comment (kobawan) [inline src/util/requiredValues.js]: typo `inLineRessources`

File: src/parser/parser_verification.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ * @param  {Boolean} [isAdInline] - True if node is contained inside a InLine node.
+ *
+ */
+function verifyRequiredValues(node, emit, isAdInline) {
+  if (!node || !node.nodeName) {
+    return;
+  }
+  if (node.nodeName === 'InLine') {
+    isAdInline = true;
```
Comment (kobawan) [inline src/parser/parser_verification.js]: why are we passing `isAdInline` as a param to the function if you're checking for it here?

File: src/parser/parser_verification.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  }
+  verifyRequiredAttributes(node, emit);
+
+  if (hasSubElements(node)) {
+    verifyRequiredSubElements(node, emit, isAdInline);
+    for (const nodeKey in node.children) {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: I'm not sure but have you checked that `for...in` doesn't loop also through keys that you don't need. AFAIK `for...in` also loops from its constructor's prototype: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in

File: src/parser/parser_verification.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+      verifyRequiredValues(node.children[nodeKey], emit, isAdInline);
+    }
+  } else if (parserUtils.parseNodeText(node).length === 0) {
+    emitMissingValueWarning({ node }, emit);
+    return;
+  }
+
+  return;
```
Comment (kobawan) [inline src/parser/parser_verification.js]: no need to return here :)

File: src/parser/parser_verification.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  verifyRequiredAttributes(node, emit);
+
+  if (hasSubElements(node)) {
+    verifyRequiredSubElements(node, emit, isAdInline);
+    for (const nodeKey in node.children) {
+      verifyRequiredValues(node.children[nodeKey], emit, isAdInline);
+    }
+  } else if (parserUtils.parseNodeText(node).length === 0) {
+    emitMissingValueWarning({ node }, emit);
+    return;
```
Comment (kobawan) [inline src/parser/parser_verification.js]: same as other comment

File: src/parser/parser_verification.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+ * @param  {Node} node - The node element
+ * @param  {Boolean} isAdInline - True if node is contained in a inline
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ */
+function verifyRequiredSubElements(node, emit, isAdInline) {
+  const required = requiredValues[node.nodeName];
+  if (!required) {
+    return;
+  }
+  const requiredSubElements = required.subElements;
```
Comment (kobawan) [inline src/parser/parser_verification.js]: this could be undefined as well, as not every requiredValues from the config has subElements mentioned

File: src/parser/parser_verification.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+/**
+ * Verify and trigger warnings if node required attributes are not set.
+ * @param  {Node} node - The node element.
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ */
+function verifyRequiredAttributes(node, emit) {
+  if (!requiredValues[node.nodeName]) {
+    return;
+  }
+  const requiredAttributes = requiredValues[node.nodeName].attributes;
```
Comment (kobawan) [inline src/parser/parser_verification.js]: this could be undefined

File: src/parser/parser_verification.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    if (!subElement) {
+      emitMissingValueWarning(
+        { node, name: subElementName, type: 'sub element' },
+        emit
+      );
+    }
+  }
+
+  // When InLine format is used some node require at least one resource (i.e <NonLinear>, <Companion>, or <Icon>)
+  if (!isAdInline || !required.inLineRessources) {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: typo!

File: src/parser/parser_verification.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    return;
+  }
+
+  let ressourceFound = false;
+  for (const ressourceKey in required.inLineRessources) {
+    const ressource = required.inLineRessources[ressourceKey];
+    if (parserUtils.childByName(node, ressource)) {
+      ressourceFound = true;
```
Comment (kobawan) [inline src/parser/parser_verification.js]: I don't understand this block, you should make sure every resource in inLineResources is present, not just the first resource found

File: src/parser/parser_verification.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+ * Check if a node is sub elements.
+ * @param  {Node} node - The node element.
+ * @returns {Boolean}
+ */
+function hasSubElements(node) {
+  if (
+    !node.children ||
+    node.children.length === 0 ||
+    (node.children.length === 1 &&
+      ['#text', '#cdata-section'].indexOf(node.children[0].nodeName) >= 0)
```
Comment (kobawan) [inline src/parser/parser_verification.js]: I don't think this line should be checked here, this function should just check if it has children. Because what happens if this element has multiple children but they all have no `#text` or `#cdata-section`?

File: src/parser/parser_verification.js [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+  return true;
+}
+
+/**
+ * Trigger Warning for if a node is empty or has missing attributes/subelements
+ * @param  {Object} missingElement - Object containing the node with empty value, or with missing attribute/subElement name and type
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ * @emits  VastParser#VAST-warning
+ */
+function emitMissingValueWarning(missingElement, emit) {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: you could already destructure `missingElement`, that way we know what this function expects

File: src/parser/parser_verification.js [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    if (parserUtils.childByName(node, ressource)) {
+      ressourceFound = true;
+      break;
+    }
+  }
+  if (!ressourceFound) {
+    emitMissingValueWarning(
+      {
+        node,
+        name: `${required.inLineRessources.join(' or ')}`,
```
Comment (kobawan) [inline src/parser/parser_verification.js]: You should clarify which resource is missing. The warning messages should be explicit

File: src/parser/parser_verification.js (hunk 14 above)
Comment (kobawan) [reply src/parser/parser_verification.js]: also maybe instead of having `{ node, name, type }` , you could pass `{ name, parentName, attributes, subElements }`. And in the message check if attributes is passed, or if subElements is passed, or neither, and make a message accordingly.

File: src/parser/ad_parser.js [hunk 16]
```diff
@@ -79,7 +77,8 @@ function parseInLine(inLineElement) {
                 case 'Linear':
                   parsedCreative = parseCreativeLinear(
                     creativeTypeElement,
-                    creativeAttributes
+                    creativeAttributes,
+                    emit
```
Comment (kobawan) [inline src/parser/ad_parser.js]: this is not needed anymore

File: src/parser/ad_parser.js [hunk 17]
```diff
@@ -326,10 +329,7 @@ export function _parseAdVerifications(verifications) {
     const verification = new AdVerification();
     const childNodes = verificationNode.childNodes;
 
-    parserUtils.assignAttributes(
-        verificationNode.attributes,
-        verification
-    );
+    parserUtils.assignAttributes(verificationNode.attributes, verification);
```
Comment (kobawan) [inline src/parser/ad_parser.js]: is this needed if you're calling it in line 42?

File: src/parser/ad_parser.js (hunk 16 above)
Comment (ZacharieTFR) [reply src/parser/ad_parser.js]: I will remove it

File: src/parser/ad_parser.js (hunk 17 above)
Comment (ZacharieTFR) [reply src/parser/ad_parser.js]: It's not part of my code, Prettier did that 😅

File: src/parser/parser_verification.js (hunk 6 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: I'll replace it with forEach

File: src/parser/parser_verification.js (hunk 10 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: I'll add the check

File: src/parser/parser_verification.js (hunk 9 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: Yep! I'll add the check

File: src/parser/parser_verification.js (hunk 12 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: As discussed in person, the iab specs require at least one of the following resources: StaticResource, IFrameResource, HTMLResource for the nodes <NonLinear>, <Companion>, or <Icon> when InLine format is used

File: src/parser/parser_verification.js (hunk 13 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: Indeed, i will remove it

File: src/parser/parser_verification.js (hunk 5 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: `isAdInline` is used to know if a node is contained inside an InLine. Since it's a recursive function, it can go through multiple childs levels. I need to pass it through recursive iteration to know when a resource requires it ( one of StaticResource, IFrameResource, HTMLResource is required for the nodes NonLinear, Companion, or Icon when InLine format is used). As discussed in person I'll explain it in the JSDoc

File: src/parser/ad_parser.js [hunk 18]
```diff
@@ -116,7 +128,8 @@ function parseInLine(inLineElement) {
 
       case 'AdVerifications':
         ad.adVerifications = _parseAdVerifications(
-          parserUtils.childrenByName(node, 'Verification')
+          parserUtils.childrenByName(node, 'Verification'),
+          emit
```
Comment (kobawan) [inline src/parser/ad_parser.js]: stray emit caught! :D

File: src/parser/ad_parser.js [hunk 19]
```diff
@@ -316,6 +330,7 @@ function _parseExtension(extNode) {
 /**
  * Parses the AdVerifications Element.
  * @param  {Array} verifications - The array of verifications to parse.
+ * @param  {Function} emit - Emit function used to trigger Warning event
```
Comment (kobawan) [inline src/parser/ad_parser.js]: no need

File: src/parser/parser_verification.js [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+function verifyRequiredAttributes(node, emit) {
+  if (
+    !requiredValues[node.nodeName] ||
+    !requiredValues[node.nodeName].attributes
+  ) {
+    return;
+  }
+  const requiredAttributes = requiredValues[node.nodeName].attributes;
+  const missingAttributes = [];
+  requiredAttributes.forEach(attributeName => {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: if you wanna be supah cool you could use `filter` method :) it is much less verbose

File: src/parser/parser_verification.js [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+
+  if (!required) {
+    return;
+  }
+
+  if (required.subElements) {
+    const requiredSubElements = required.subElements;
+    const missingSubElements = [];
+
+    requiredSubElements.forEach(subElementName => {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: same here with `filter`

File: src/parser/parser_verification.js [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+    }
+  }
+
+  // When InLine format is used some nodes (i.e <NonLinear>, <Companion>, or <Icon>)
+  // require at least one of the following resources: StaticResource, IFrameResource, HTMLResource
+  if (!isAdInline || !required.inLineResources) {
+    return;
+  }
+
+  const resourceFound = required.inLineResources.some(resource => {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: are all the InLine subelements here optional, like `StaticResource, IFrameResource, HTMLResource`?

File: src/parser/parser_verification.js [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+  }
+}
+
+/**
+ * Check if a node has sub elements.
+ * @param  {Node} node - The node element.
+ * @returns {Boolean}
+ */
+function hasSubElements(node) {
+  return !node.children || node.children.length !== 0;
```
Comment (kobawan) [inline src/parser/parser_verification.js]: shouldn't it be the opposite? `return node.children && node.children.length`

File: src/parser/parser_verification.js [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+    isAdInline = true;
+  }
+  verifyRequiredAttributes(node, emit);
+
+  if (hasSubElements(node)) {
+    verifyRequiredSubElements(node, emit, isAdInline);
+    for (const nodeKey in node.children) {
+      verifyRequiredValues(node.children[nodeKey], emit, isAdInline);
+    }
+  } else if (parserUtils.parseNodeText(node).length === 0) {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: does this take only cdata or text? We don't want comments

File: src/parser/parser_verification.js (hunk 22 above)
Comment (kobawan) [reply src/parser/parser_verification.js]: as discussed in person, we shoud rename `inLineResources` to `oneOfinLineResources`

File: src/parser/parser_verification.js (hunk 23 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: Totally ! I'll will fix it right away!

File: src/parser/parser_verification.js (hunk 24 above)
Comment (kobawan) [reply src/parser/parser_verification.js]: ah great to know!

File: src/parser/ad_parser.js [hunk 25]
```diff
@@ -27,29 +29,42 @@ export function parseAd(adElement) {
...
+ * Parses an ad type (Inline or Wrapper)
+ * @param  {Object} adTypeElement - The VAST Inline or Wrapper element to parse.
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ * @emits  VASTParser#VAST-warning
  * @return {Ad}
  */
-function parseInLine(inLineElement) {
-  const childNodes = inLineElement.childNodes;
+function parseAdElement(adTypeElement, emit) {
+  parserVerification.verifyRequiredValues(adTypeElement, emit);
```
Comment (kobawan) [inline src/parser/ad_parser.js]: you should do the check for no `emit` here, so you don't have to call this function

File: src/parser/parser_verification.js [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+ * @param  {Array} missingElement.subElements - The array of missing sub elements
+ * @param  {Array} missingElement.oneOfResources - The array of resources in which at least one must be provided by the element
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ * @emits  VastParser#VAST-warning
+ */
+function emitMissingValueWarning(
+  { name, parentName, attributes, subElements, oneOfResources },
+  emit
+) {
+  if (!emit) {
```
Comment (kobawan) [inline src/parser/parser_verification.js]: as mentioned check should be done way before-hand

File: src/parser/parser_verification.js (hunk 6 above)
Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: Actually it's not an array it's an HTMLCollection so i can't use a forEach 😞 . I'll  replace for/in with the classical for loop.

File: spec/vast_verification.spec.js [hunk 27]
```diff
@@ -0,0 +1,122 @@
+import { parserVerification } from '../src/parser/parser_verification';
+import { vastInLine } from './samples/inline_missing_values';
+import { vastMissingWrapperValues } from './samples/wrapper_missing_values';
+
+describe('parserVerification', function() {
+  describe('verifyRequiredValues', function() {
+    let warnings = [];
+
+    const parseXmlFromString = xmlString => {
```
Comment (kobawan) [inline spec/vast_verification.spec.js]: I would put this under `spec/utils/utils.js`

File: spec/vast_verification.spec.js [hunk 28]
```diff
@@ -0,0 +1,122 @@
+import { parserVerification } from '../src/parser/parser_verification';
+import { vastInLine } from './samples/inline_missing_values';
+import { vastMissingWrapperValues } from './samples/wrapper_missing_values';
+
+describe('parserVerification', function() {
```
Comment (kobawan) [inline spec/vast_verification.spec.js]: Right now you tested that you get warnings for all the things you needed, but you didn't add any tests to check that `verifyRequiredValues`, `verifyRequiredAttributes `, `verifyRequiredSubElements `, `hasSubElements `, `emitMissingValueWarning ` individually do what you expect them to do.

File: spec/vast_verification.spec.js (hunk 27 above)
Comment (ZacharieTFR) [reply spec/vast_verification.spec.js]: I removed it, and replace it with the one already in the utils

File: spec/vast_verification.spec.js [hunk 29]
```diff
@@ -0,0 +1,254 @@
...
+      });
+    });
+  });
+
+  describe('verifyRequiredAttributes', function() {
+    it('should call the emit function for missing required attributes', () => {
+      const emitMock = jest.fn();
+      const element = getNodesFromXml('<MediaFile></MediaFile>');
+      parserVerification.verifyRequiredAttributes(element, emitMock);
+      expect(emitMock.mock.calls.length).toBe(1);
```
Comment (kobawan) [inline spec/vast_verification.spec.js]: it is best to use native jest fn `toHaveBeenCalledTimes`

File: src/parser/parser_verification.js [hunk 30]
```diff
@@ -69,8 +69,13 @@ function verifyRequiredAttributes(node, emit) {
  */
 function verifyRequiredSubElements(node, emit, isAdInline) {
   const required = requiredValues[node.nodeName];
-
-  if (!required) {
+  // Wrapper child elements have no required subElement. (Only InLine does)
+  if (
+    !required ||
+    (!isAdInline &&
```
Comment (kobawan) [inline src/parser/parser_verification.js]: you should only depend on `isAdInline` here. It is confusing that you have to check for the node name.

Comment (ZacharieTFR) [reply src/parser/parser_verification.js]: As discussed i will use a better naming 😉

File: src/parser/parser_verification.js [hunk 31]
```diff
@@ -0,0 +1,179 @@
...
+ * @param  {Array} missingElement.subElements - The array of missing sub elements
+ * @param  {Array} missingElement.oneOfResources - The array of resources in which at least one must be provided by the element
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ * @emits  VastParser#VAST-warning
+ */
+function emitMissingValueWarning(
+  { name, parentName, attributes, subElements, oneOfResources },
+  emit
+) {
+  let message = `Element `;
```
Comment (clementFrancon) [inline src/parser/parser_verification.js]: Might be better to put the name here already and use else if,  and else for the last statement, if we can only have one kind of error per message.
